src/js/service/PiskelApiService.js

```diff
@@ -0,0 +1,157 @@
+    this.allowedOrigin_ = piskelWindow.location.origin;
+    piskelWindow.addEventListener('message', this.receiveMessage_.bind(this));
+
+    // For implementing auto-save: On a save state event, we should notify
+    // the parent app that the animation has changed.
+    $.subscribe(Events.PISKEL_SAVE_STATE, this.onSaveStateEvent.bind(this));
+
+    this.log('Initialized.');
+  };
+
+  // Debug logging utility - should remove at some point.
+  var VERBOSE = true;
+  ns.PiskelApiService.prototype.log = function () {
+    if (VERBOSE) {
+      var toLog = ['PiskelApiService:'];
+      for (var i = 0; i < arguments.length; i++) {
+        toLog.push(arguments[i]);
+      }
+      console.log.apply(console, toLog);
+    }
+  };
+
+  /**
+   * Send a cross-context message out to the parent window.
+   * @param {Object} message
+   * @private
+   */
+  ns.PiskelApiService.prototype.sendMessage_ = function (message) {
+    this.parentWindow_.postMessage(message, this.allowedOrigin_);
+  };
+
+  /**
+   * Entry point for routing messages received from the Code.org parent document.
+   * @param {Event} event
+   * @private
+   */
+  ns.PiskelApiService.prototype.receiveMessage_ = function (event) {
+    // Ignore messages not sent from the allowed origin
+    var origin = event.origin || event.originalEvent.origin;
+    if (origin !== this.allowedOrigin_) {
+      return;
+    }
+
+    var message = event.data;
+    this.log('Received message:', message);
+
+    // Delegate according to message type
+    if (message.type === MessageType.LOAD_SPRITESHEET) {
+      this.loadSpritesheet(message.uri, message.frameSizeX, message.frameSizeY);
+    }
+  };
+
+  ns.PiskelApiService.prototype.loadSpritesheet = function (uri, frameSizeX, frameSizeY) {
+    var image = new Image();
+    image.onload = function () {
+      // Avoid retriggering image onload (something about JsGif?)
+      image.onload = function () {};
+      this.importService_.newPiskelFromImage(image, {
+        importType: 'spritesheet',
+        frameSizeX: frameSizeX,
+        frameSizeY: frameSizeY,
+        frameOffsetX: 0,
+        frameOffsetY: 0,
+        smoothing: false
+      }, function () {
+        this.log('Image loaded.');
+        // TODO: Report load complete out to parent app?
+      }.bind(this));
+    }.bind(this);
+    image.src = uri;
+  };
+
+  ns.PiskelApiService.prototype.onSaveStateEvent = function (evt, action) {
+    this.log('onSaveStateEvent');
+    // Render spritesheet and return blob, dataURI and metadata to parent app.
+    var renderer = new pskl.rendering.PiskelRenderer(this.piskelController_);
+    var outputCanvas = renderer.renderAsCanvas(this.getBestFit_());
+    var dataURI = outputCanvas.toDataURL();
+    pskl.utils.BlobUtils.dataToBlob(dataURI, 'image/png', function (blob) {
+      this.sendMessage_({
+        type: MessageType.STATE_SAVED,
+        blob: blob,
+        dataURI: dataURI,
```

Comment:
I'm sending both the blob and dataURI here because it seems to be a pretty quick conversion and it's convenient to have both on the other end - the dataURI to push directly into the animation previews in the left column, and the blob for upload to S3.  If we think it's cleaner I can send only the blob, as a Blob-to-dataURI conversion is pretty straightforward with the standard [`FileReader` object](https://developer.mozilla.org/en-US/docs/Web/API/FileReader).  Going the other way (dataURI-to-Blob) is messier; a stackoverflow search revealed lots of (different) homemade solutions, but no standard approach.  At least by doing the conversion here, we know Piskel is doing things in its own internally-consistent way.